Implement hold on channels

When we online a child we need a reference to the IO channels. Right now this is implemented as:

```rust
let ch = get_io_channel()
... code
put_io_channel(ch)
```

It would be nice if we can do a simple `self.channel.hold()` or something like that

Thanks for the tips! I've managed to get things almost working with the docker image. I had to update the rust toolchain to get some dependencies to compile, but I'm getting a link error on mayastor:

> error: linking with `cc` failed: exit code: 1
>   |
>   = note: /usr/bin/ld: /root/MayaStor/target/debug/deps/libmayastor-96c3dba6c179a74d.rlib(mayastor-96c3dba6c179a74d.4p
> i9smpals94548h.rcgu.o): in function `mayastor::bdev::nexus::nexus_share::<impl mayastor::bdev::nexus::nexus_bdev::Nexu
> s>::share::{{closure}}':
>           /root/MayaStor/mayastor/src/bdev/nexus/nexus_share.rs:45: undefined reference to `create_crypto_disk'
>           /usr/bin/ld: /root/MayaStor/target/debug/deps/libmayastor-96c3dba6c179a74d.rlib(mayastor-96c3dba6c179a74d.4p
> i9smpals94548h.rcgu.o): in function `mayastor::bdev::nexus::nexus_share::<impl mayastor::bdev::nexus::nexus_bdev::Nexu
> s>::unshare::{{closure}}':
>           /root/MayaStor/mayastor/src/bdev/nexus/nexus_share.rs:92: undefined reference to `delete_crypto_disk'
>           collect2: error: ld returned 1 exit status

But i don't think we have to solve this immediately, as compiling is enough for this matter. I'll send an initial version of my patch for review. Thanks for your attention by the way.

On Wed, Oct 23, 2019 at 8:08 PM gahag <notifications@github.com> wrote:

> Thanks for the tips! I've managed to get things almost working with the
> docker image. I had to update the rust toolchain to get some dependencies
> to compile, but I'm getting a link error on mayastor:
>
> error: linking with cc failed: exit code: 1
> |
> = note: /usr/bin/ld:
> /root/MayaStor/target/debug/deps/libmayastor-96c3dba6c179a74d.rlib(mayastor-96c3dba6c179a74d.4p
> i9smpals94548h.rcgu.o): in function mayastor::bdev::nexus::nexus_share::<impl
> mayastor::bdev::nexus::nexus_bdev::Nexu s>::share::{{closure}}':
> /root/MayaStor/mayastor/src/bdev/nexus/nexus_share.rs:45: undefined
> reference to create_crypto_disk'
> /usr/bin/ld:
> /root/MayaStor/target/debug/deps/libmayastor-96c3dba6c179a74d.rlib(mayastor-96c3dba6c179a74d.4p
> i9smpals94548h.rcgu.o): in function mayastor::bdev::nexus::nexus_share::<impl
> mayastor::bdev::nexus::nexus_bdev::Nexu s>::unshare::{{closure}}':
> /root/MayaStor/mayastor/src/bdev/nexus/nexus_share.rs:92: undefined
> reference to delete_crypto_disk'
> collect2: error: ld returned 1 exit status
>
>  I've actually just updated the docs because it indeed is something that
requires some explanation! The container image comes pre-packed with a
library (libspdkfat.so) which is somewhat outdated now. To get it up to
date, within the container:

cd spdk-sys
git submodule update --init --recursive
./build.sh --enable-debug --without-isal --with-iscsi-initiator --with-rdma
\
             --with-internal-vhost-lib --disable-tests \
             --with-crypto
cp build/libspdk_fat.so /lib

Then it should work :)


> But i don't think we have to solve this immediately, as compiling is
> enough for this matter. I'll send an initial version of my patch for
> review. Thanks for your attention by the way.
>

My pleasure and thanks to you too!

> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openebs/MayaStor/issues/9?email_source=notifications&email_token=ADZX7VIZLVZIKXEOIBMJUXDQQCHJRA5CNFSM4I6CNF72YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECCLK4I#issuecomment-545568113>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADZX7VKGD4GLC3OC2G3UKQDQQCHJRANCNFSM4I6CNF7Q>
> .
>

--- a/mayastor/src/bdev/nexus/nexus_bdev_children.rs
+++ b/mayastor/src/bdev/nexus/nexus_bdev_children.rs
@@ -3,6 +3,8 @@ use crate::bdev::{
     nexus::{self, nexus_child::NexusChild, Error},
 };
 
+use crate::descriptor::IoChannel;
+
 use crate::nexus_uri::BdevType;
 
 use spdk_sys::{spdk_get_io_channel, spdk_put_io_channel};
@@ -17,6 +19,10 @@ use crate::bdev::nexus::{
 use futures::future::join_all;
 
 impl Nexus {
+    fn hold_io_channel(&self) -> IoChannel {
+        IoChannel::new(self.as_ptr())
+    }
+
     /// Add the child bdevs to the nexus instance in the "init state"
     /// this function should be used when bdevs are added asynchronously
     /// like for example, when parsing the init file. The examine callback
@@ -72,9 +78,10 @@ impl Nexus {
 
         if let Some(child) = self.children.iter_mut().find(|c| c.name == name) {
             child.close()?;
-            let ch = unsafe { spdk_get_io_channel(self.as_ptr()) };
-            self.reconfigure(DREvent::ChildOffline).await;
-            unsafe { spdk_put_io_channel(ch) }
+            {
+                let _ch = self.hold_io_channel();
+                self.reconfigure(DREvent::ChildOffline).await;
+            }
             self.set_state(NexusState::Degraded);
             Ok(NexusState::Degraded)
         } else {

--- a/mayastor/src/descriptor/mod.rs
+++ b/mayastor/src/descriptor/mod.rs
@@ -47,11 +47,13 @@ use spdk_sys::{
     spdk_bdev_write,
     spdk_dma_free,
     spdk_dma_zmalloc,
+    spdk_get_io_channel,
     spdk_io_channel,
     spdk_put_io_channel,
 };
 use std::{
     ffi::c_void,
+    ops::{Deref, DerefMut},
     slice::{from_raw_parts, from_raw_parts_mut},
 };
 
@@ -331,5 +333,46 @@ impl Drop for Descriptor {
                 spdk_bdev_close(self.desc)
             }
         };
+    }
+}
+
+/// Automatic IO channel handle
+#[derive(Debug)]
+pub struct IoChannel {
+    /// the channel handle
+    handle: *mut spdk_io_channel,
+}
+
+impl IoChannel {
+    /// Acquire an io channel for the given nexus.
+    /// This channel has guard semantics, and will be released when dropped.
+    pub fn new(nexus: *mut c_void) -> Self {
+        unsafe {
+            IoChannel {
+                handle: spdk_get_io_channel(nexus),
+            }
+        }
+    }
+}
+
+impl Drop for IoChannel {
+    fn drop(&mut self) {
+        unsafe {
+            spdk_put_io_channel(self.handle);
+        };
+    }
+}
+
+impl Deref for IoChannel {
+    type Target = *mut spdk_io_channel;
+
+    fn deref(&self) -> &Self::Target {
+        &self.handle
+    }
+}
+
+impl DerefMut for IoChannel {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.handle
     }
 }

